drivers/disk/ramdisk.c

```diff
@@ -142,6 +142,23 @@ static const struct disk_operations ram_disk_ops = {
 		.buf = UINT_TO_POINTER(DT_REG_ADDR(DT_INST_PHANDLE(n, ram_region))),	\
 	}
 
+#define RAMDISK_DEVICE_CONFIG_DEFINE_DISKIMG(n)                                         \
+__asm__(                                                                                \
+"	.section   .data\n"                                                             \
+"	.balign    8\n"                                                                 \
+"	.global    disk_buf_" STRINGIFY(n) "\n"                                         \
+"disk_buf_" STRINGIFY(n) ":\n"                                                          \
+"	.incbin    \"" DT_INST_PROP(n, disk_image) "\"\n");                             \
+											\
+	extern uint8_t disk_buf_##n[] __asm__("disk_buf_" STRINGIFY(n));                \
+											\
+	static struct ram_disk_config disk_config_##n = {				\
+		.sector_size = DT_INST_PROP(n, sector_size),				\
+		.sector_count = DT_INST_PROP(n, sector_count),				\
+		.size = RAMDISK_DEVICE_SIZE(n),						\
+		.buf = disk_buf_##n,							\
+	}
+
```

Comment:
>Note that this implementation uses .incbin directive which is technically
GNU-specific but is widely supported by other mainstream toolchains as well
(for compatibility with GCC). C23 #embed directive can be a more portable
option but toolchains supporting it are just coming out

I see no reasons to use something GNU-specific here or justify it with something that could happen in the distant future. There also no need to introduce this linker script upstream, please define memory region in your application and use phandle ram-region, `ram-region = <&my_fancy_binary_image>`.

Reply:
I already considered that but ram-region property does _not_ solve the problem cleanly:
* Defining a ram-region is an intrusive change to a platform device-tree. You'll have to take memory out of your other existing memory regions, which makes it highly platform (board) specific. Implementation will also be comparatively ugly as you'd need board-specific device-tree overlays to achieve it.
* If disk image size changes during development, you may need to update the device-tree modification that was made to define the custom ram-region. Or you'll need to give it some generous size to account for changes in disk size (a liberty you won't have if RAM is not aplenty).
* You'll also need an application specific linker script which will be more GNU-specific than .incbin.
* And then you may have to separately map the region to make it accessible (HW dependency).

(Let me know if I misunderstood something and there can be a simpler/cleaner way to use ram-region property for the usecase being addressed here).

With the proposed change-set, all of this gets handled in a transparent manner. All you need is a (generic) device-tree overlay, and you are set. The embedded disk image will get placed in the data region and will automatically get adjusted even if its size changes - nothing extra needs to be done. And when you don't need to embed the disk image anymore, simply remove the overlay (or remove the disk-image-path property, if still using that RAM disk instance) and rebuild.
```
	ramdisk0 {
		compatible = "zephyr,ram-disk";
		disk-name = "RAM";
		sector-size = <512>;
		sector-count = <0>;
		zephyr,disk-image-path = "../ramdisk.img";
	};
```
Note: `sector-count` will be automatically calculated from the `sector-size` and size of the disk image being embedded.

IMHO, the present enhancement is orthogonal to ram-region property (which can be more suitable for the usecases it was originally introduced for, not the one addressed here) so would be a nice enhancement to out-of-box features provided by Zephyr.

Regarding portability concerns, .incbin is already fairly portable. The feature itself is widely available, only the syntax may vary (if you move away from major toolchains). Still if this is a major concern, we can consider using a portable wrapper **incbin.h** (https://github.com/graphitemaster/incbin). IMHO though, given the toolchains that would be typically used with Zephyr, that would be a bit overkill.